README.md
# MVR
Mean-Variance Regularization: a non-parametric method for joint adaptive mean-variance regularization and variance stabilization of high-dimensional data


===============
### Description

Implements a non-parametric method for joint adaptive mean-variance regularization and variance stabilization of 
high-dimensional data (Dazard and Rao, 2012). It is suited for handling difficult problems posed by high-dimensional 
multivariate datasets (_p_ >> _n_ paradigm), such as in omics-type data, among which are that the variance is often 
a function of the mean, variable-specific estimators of variances are not reliable, and tests statistics have low powers 
due to a lack of degrees of freedom.

Key features include:

1. Normalization and/or variance stabilization of the data

2. Computation of mean-variance-regularized _t_-statistics (_F_-statistics to come)

3. Generation of diverse diagnostic plots

4. Computationally efficient implementation using C/C++ interfacing and an option for parallel
computing to enjoy a fast and easy experience in the R environment

See also below the package news with the R command: `MVR.news()`.

All the codes are in the R folder and a manual (MVR.pdf) details the end-user (and internal) functions. 
At this stage and for simplicity, there are only 2 end-user function, 4 end-user diagnostic 
and plotting functions and 2 end-user datasets (synthetic and real). 
See the "MVR-package" introduction section of the manual for more details and examples.


============
### Branches

This branch (master) is the  default one, that hosts the current development release (version 1.34.0).

===========
### License

MVR is open source / free software, licensed under the GNU General Public License version 3 (GPLv3), 
sponsored by the [Free Software Foundation](https://www.fsf.org/). To view a copy of this license, visit 
[GNU Free Documentation License](https://www.gnu.org/licenses/gpl-3.0.html).


=============
### Downloads

CRAN downloads since October 1, 2012, 
the month the [RStudio CRAN mirror](http://cran-logs.rstudio.com/) 
started publishing logs:
[![](https://cranlogs.r-pkg.org/badges/grand-total/MVR)](https://CRAN.R-project.org/package=MVR)

CRAN downloads in the last month:
[![](https://cranlogs.r-pkg.org/badges/last-month/MVR)](https://CRAN.R-project.org/package=MVR)

CRAN downloads in the last week:
[![](https://cranlogs.r-pkg.org/badges/last-week/MVR)](https://CRAN.R-project.org/package=MVR)


================
### Requirements

MVR (>= 1.34.0) requires R-3.5.0 (2018-04-23). It was built and tested under R version 4.0.3 (2020-10-10) and Travis CI. 

Installation has been tested on Windows, Linux, OSX and Solaris platforms. 

See Travis CI build result:
[![Build Status](https://travis-ci.com/jedazard/MVR.svg)](https://travis-ci.com/jedazard/MVR)

See CRAN checks:
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/MVR)](https://cran.r-project.org/web/checks/check_results_MVR.html)


================
### Installation

* To install the stable version of `MVR`, simply download and install the current version (1.33.0) from the [CRAN](https://CRAN.R-project.org/package=MVR) 
repository:

```{r}
install.packages("MVR")
```

* Alternatively, you can install the most up-to-date development version (>= 1.34.0) of `MVR` from the [GitHub](https://github.com/jedazard/MVR) repository:

```{r}
install.packages("devtools")
library("devtools")
devtools::install_github("jedazard/MVR")
```

=========
### Usage

* To load the MVR library in an R session and start using it:

```{r}
library("MVR")
```

* Check the package news with the R command:

```{r}
MVR.news()
```

* Check on how to cite the package with the R command:

```{r}
citation("MVR")
```

etc...


===================
### Acknowledgments

Authors: 
   + Jean-Eudes Dazard, Ph.D. <jean-eudes.dazard@case.edu>
   + Hua Xu, Ph.D. <huaxu77@gmail.com>
   + Alberto Santana, MBA. <ahs4@case.edu>

Maintainers: 
   + Jean-Eudes Dazard, Ph.D. <jean-eudes.dazard@case.edu>

Funding/Provision/Help:   
   + This work made use of the High Performance Computing Resource in the Core Facility for Advanced Research Computing at Case Western Reserve University. 
   + This project was partially funded by the National Institutes of Health NIH - National Cancer Institute (P30-CA043703).


==============
### References
              
   + Dazard J-E. and J. S. Rao.
      *Joint Adaptive Mean-Variance Regularization and Variance Stabilization of High Dimensional Data*. 
      [Comput. Statist. Data Anal. (2012)](https://www.sciencedirect.com/science/article/abs/pii/S0167947312000321), 56(7):2317-2333.
      (The Official Journal of the International Association for Statistical Computing).

   + Dazard J-E., Hua Xu and J. S. Rao. 
      *R package MVR for Joint Adaptive Mean-Variance Regularization and Variance Stabilization*. 
      In JSM Proceedings, Section for Statistical Programmers and Analysts. 
      Miami Beach, FL, USA: American Statistical Association IMS - JSM, 3849-3863.
      [JSM (2011)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4725579/).

   + Dazard J-E. and J. S. Rao. 
      *Regularized Variance Estimation and Variance Stabilization of High-Dimensional Data*. 
      In JSM Proceedings, Section for High-Dimensional Data Analysis and Variable Selection. 
      Vancouver, BC, Canada: American Statistical Association IMS - JSM, 5295-5309.
      [JSM (2010)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4727967/).


inst/NEWS
Package: MVR

---------------------------------------------------------------------------------
Date   : 2011-07-26
o RELEASE 1.00.0
- Initial release of the package.
---------------------------------------------------------------------------------
Date   : 2011-12-14
o RELEASE 1.10.0
- Minor fixes in the documentation.
- The synthetic dataset is downsized to 200 variables (instead of 1000) with 40 of them (instead of 200) that are significantly different between the two sample groups.
- Major bug fix in the internal function MeanVarReg, which was not returning the correct cluster membership values. This was affecting especially the end-user function mvrt.test().
---------------------------------------------------------------------------------
Date   : 2013-11-13
o RELEASE 1.20.0
- Now depends on 'parallel' instead of 'snow' package. This modification should be totally transparent to the users and not affect them in any way. Minor updates in the manual accordingly.
- Now supports SOCKET, PVM and MPI clusters. Manual corrected accordingly.
- Minor updates in the email addresses and references.
---------------------------------------------------------------------------------
Date   : 2015-06-19
o RELEASE 1.30.1
- Packages in 'Depends' field of the DESCRIPTION file are now imported from the NAMESPACE file.
- Compiled C code now uses the interface to R internal RNGs routines instead of the system rand RNG.
- Compiled C code now successfully passes Valgrind memory check. 
- Removed superfluous codes in the parallelization section, before clusterCall().
- Main function `mvr()` now returns shrinkage estimates of means and standard deviations by groups. 
- Improved codes for all plotting functions, including new graphical device arguments.
- Updates in the manual: examples are given for a SOCKET or MPI cluster configuration.
- Removed internal functions from the manual, updated manual, version number.
- Test build under 3.0.2 and release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2015-07-10
o RELEASE 1.30.2
- Compliance with new R CMD check, which now checks code usage via 'codetools'.
  Functions and packages from default packages other than base which are used in the package
  code are now imported via the package namespace file (NAMESPACE).
  Added new field 'Imports' in the package description file (DESCRIPTION) 
  to match the functions and packages newly imported via NAMESPACE.
- Minor updates of the default title values in the diagnostic plots. 
- Built under R-3.2.1 (2015-06-19) and tested under R-devel (2015-07-08 r68642). 
- Release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2016-01-28
o RELEASE 1.30.3
- Corrected inclusion of C and R headers to comply with g++ 6.0 compiler and good practice in C++ code.
- Added manual in a 'doc' subfolder within 'inst' folder.
- Built and tested under R-devel (2015-12-22 r69809). 
- Release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2016-10-25
o RELEASE 1.31.0
BUG FIXES
- Corrected missing declaration of use of 'std' namespace in C++ code to comply with C++98 and good practice.
- Corrected number of replications to match the exact same number as user request and not a multiple of the number of cpus available.
UTILITIES
- Added a seed for both end-user functions `mvr()` and `mvrt.test()`.
PACKAGE INSTALLATION
- Updated Rd, NEWS and CITATION files.  
- Built and tested under R-devel (2016-06-30 r70858) and release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2017-05-28
o RELEASE 1.32.0
- Minor corrections in R and C/C++ codes to comply with good practices of
  native routines registration and symbol search disabling.
PACKAGE INSTALLATION
- Updated documentation, Rd, DESCRIPTION, NEWS and NAMESPACE files.  
- Built and tested under R version 3.4.0 (2017-04-21) and release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2018-04-16
o RELEASE 1.32.1
BUG FIXES
- NA values are now correctly handled in the internal function `is.empty()`.
- a NULL value for the seed is now correctly handled in case of no parallelization 
  in the internal functions `withinsumsq()` and `sim.dis()`.
PACKAGE INSTALLATION
- Minor updates in documentation files (DESCRIPTION, NEWS and README).
- Built and tested under R version 3.4.4 (2018-03-15) and release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2018-09-10
o RELEASE 1.33.0
BUG FIXES
- Argument `block` now correctly handles the case of a factor when the option `labels`
  or `levels` is not used to define the factor, which could cause its levels to not 
  match the data groups.
PACKAGE INSTALLATION
- Minor updates in manual and documentation files (Rd files).
- Minor updates in documentation files (DESCRIPTION, NEWS, and README).
- Built and tested under R version 3.5.1 (2018-07-02) and release update to GitHub.
---------------------------------------------------------------------------------
Date   : 2020-10-23
o RELEASE 1.34.0
SIGNIFICANT USER-VISIBLE CHANGES
- Renamed objects returned by end-user functions `mvr()` and `mvrt.test()` as `MVR` and `MVRT`, respectively.
- Corrected and updated Import and Depends field values of DESCRIPTION file.
- Removed deprecated Lazyload and LazyData field values of DESCRIPTION file.
BUG FIXES
- Conditional tests in internal function `is.empty()` no longer coerce a Boolean vector
  (logical of length > 1) to its first entry only (now passes R >= 3.6.0 CRAN check). 
- Argument `block` now correctly handles the case of a factor when the option `labels` or `levels` is not used to define the factor, which could cause its levels to not match the data groups.

src/MVRc.c
#include "MVRc.h"

//============//
// MVR_stat.c //
//============//
double sum(double* a, int n)
{
    double s = 0.0;
    for (int i = 0; i < n; ++i) {
        s += a[i];
    }
    return s;
}

double sd(double* a, double mean, int n)
{
    double sse = 0.0;
    for (int i = 0; i < n; ++i) {
        double e = a[i] - mean;
        sse += e*e;
    }
    return sqrt(sse/(double)n);
}

//============//
// MVR_rand.c //
//============//
#ifdef __MVR_C_R__
int randi()
{
    return (int)(unif_rand()*RANDF_CONST);
}

double randd()
{
    return unif_rand();
}

void MVR_rand_init()
{
    GetRNGstate();
}

void MVR_rand_end()
{
    PutRNGstate();
}

#else
int randi()
{
    return (rand()&0x7fff) | ((rand()&0x7fff) << 15);
}

//double randf()
//{
//    return (double)(randi())/RANDF_CONST;
//}

double randd()
{
    return (double)(randi())/RANDD_CONST1 + (double)(randi())/RANDD_CONST2;
}

void MVR_rand_init()
{
    srand(time(NULL));
}

void MVR_rand_end() {}
#endif

void rand_unif(double* data, int n)
{
    for (int i = 0; i < n; ++i) {
        data[i] = randd();
    }
}

//n must be even, otherwise memory overflows
void rand_norm_std(double* data, int n)
{
    double x1, x2, w;
    
    for (int i = 0; i < n; i += 2) {
        do {
            x1 = 2.0 * randd() - 1.0;
            x2 = 2.0 * randd() - 1.0;
            w = x1 * x1 + x2 * x2;
        } while ( w >= 1.0 );
        
        w = sqrt( (-2.0 * log( w ) ) / w );
        data[i] = (double)(x1 * w);
        data[i+1] = (double)(x2 * w);
    }
}

//n must be even, otherwise memory overflows
void rand_norm(double* data, int n, double mean, double s)
{
    double x1, x2, w;
    
    for (int i = 0; i < n; i += 2) {
        do {
            x1 = 2.0 * randd() - 1.0;
            x2 = 2.0 * randd() - 1.0;
            w = x1 * x1 + x2 * x2;
        } while ( w >= 1.0 );
        
        w = sqrt( (-2.0 * log( w ) ) / w ) * s;
        data[i] = (double)(x1 * w) + mean;
        data[i+1] = (double)(x2 * w) + mean;
    }
}

void rand_spl_row(double* data, int nr, int nc, double* spl, int s, int* pool)
{
    Rboolean flag_size = FALSE;
    if (s > (nr >> 1)) {
        s = nr - s;
        flag_size = TRUE;
    }
    
    //random int in [0, nr), size of s
    for (int i = 0; i < nr; ++i) pool[i] = i;
    int pool_size = nr;
    
    for (int i = 0; i < s; ++i) {
        int i_spl = randi()%pool_size;
        --pool_size;
        int tmp = pool[i_spl];
        pool[i_spl] = pool[pool_size];
        pool[pool_size] = tmp;
    }
    
    //if flag_size, keep unselected: 0 ~ pool_size-1
    //if not, keep selected: pool_size ~ nr-1
    int ir0 = pool_size,
        ir_end = nr;
    if (flag_size) {
        ir0 = 0;
        ir_end = pool_size;
        s = nr-s;
    }
    int r_des = 0;
    for (int ir = ir0; ir < ir_end; ++ir) {
        int r_src = pool[ir];
        for (int c = 0; c < nc; ++c) {
            spl[r_des+c*s] = data[r_src+c*nr];
        }
        ++r_des;
    }
}

void rand_spl_row2(double* data, int nr, int nc, double* spl, int s)
{
    int* pool = (int*) malloc(nr*sizeof(int));
    rand_spl_row(data, nr, nc, spl, s, pool);
    free(pool);
}

//==============//
// MVR_kmeans.c //
//==============//
/*
Rewritten based on kmeans_MacQueen from R distribution.
*/
Rboolean MVR_kmeans_MacQueen(double *x, double *cen,
                             int *cl, int *nc, double *wss,
                             int n, int p, int k,
                             int maxiter)
{
    /* first assign each point to the nearest cluster centre */
    int inew = 0;
    int i, j, c;
    for(i = 0; i < n; i++) {
        double best = R_PosInf;
        for(j = 0; j < k; j++) {
            double dd = 0.0;
            for(c = 0; c < p; c++) {
                double tmp = x[i+n*c] - cen[j+k*c];
                dd += tmp * tmp;
            }
            if(dd < best) {
                best = dd;
                inew = j+1;
            }
        }
        cl[i] = inew;//if(cl[i] != inew) cl[i] = inew;
    }
   /* and recompute centres as centroids */
    for(j = 0; j < k*p; j++) cen[j] = 0.0;
    for(j = 0; j < k; j++) nc[j] = 0;
    for(i = 0; i < n; i++) {
        int it = cl[i] - 1; nc[it]++;
        for(c = 0; c < p; c++) cen[it+c*k] += x[i+c*n];
    }
    for(j = 0; j < k*p; j++) cen[j] /= nc[j % k];
    
    Rboolean flag_conv = FALSE;
    for(int iter = 0; iter < maxiter; iter++) {
        flag_conv = TRUE;
        for(i = 0; i < n; i++) {
            double best = R_PosInf;
            for(j = 0; j < k; j++) {
                double dd = 0.0;
                for(c = 0; c < p; c++) {
                    double tmp = x[i+n*c] - cen[j+k*c];
                    dd += tmp * tmp;
                }
                if(dd < best) {
                    best = dd;
                    inew = j;
                }
            }
            
            int iold = cl[i] - 1;
            if(iold != inew) {
                flag_conv = FALSE;
                cl[i] = inew + 1;
                nc[iold]--; nc[inew]++;
                /* update old and new cluster centres */
                for(c = 0; c < p; c++) {
                    cen[iold+k*c] += (cen[iold+k*c] - x[i+n*c])/nc[iold];
                    cen[inew+k*c] += (x[i+n*c] - cen[inew+k*c])/nc[inew];
                }
            }
        }
        
        if(flag_conv) break;
    }
    
    for(j = 0; j < k; j++) wss[j] = 0.0;
    for(i = 0; i < n; i++) {
        int it = cl[i] - 1;
        for(c = 0; c < p; c++) {
            double tmp = x[i+n*c] - cen[it+k*c];
            wss[it] += tmp * tmp;
        }
    }
    
    return flag_conv;
}

//===========//
// MVR_sub.c //
//===========//
void MVR_km_clustering(double* x,
                       double* x_unq,
                       
                       double* centers,
                       int* cl,//vec of m
                       int* nc,//vec of k
                       double* wss,//vec of k
                       double* tot_wss,
                       int* perror,
                       
                       int* pm,
                       int* pmm,
                       int* pp,
                       int* pk,
                       int* pnstart,
                       int* pmaxiter)
{
    int m = *pm,
        mm = *pmm,
        p = *pp,
        k = *pk,
        nstart = *pnstart,
        maxiter = *pmaxiter,
        kxp = k*p;
    
    MVR_rand_init();
    
    int* rand_spl_pool = (int*) malloc(mm*sizeof(int));
    
    rand_spl_row(x_unq, mm, p, centers, k, rand_spl_pool);
    
    Rboolean flag_conv = MVR_kmeans_MacQueen(x, centers,
                                             cl, nc, wss,
                                             m, p, k,
                                             maxiter);
    *perror = flag_conv ? 0 : 1;
    double best = sum(wss, k);
    
    if (nstart > 1) {
        double *best_cen = centers,
               *best_wss = wss;
        int *best_cl = cl,
            *best_nc = nc;
        
        double* centers2 = (double*) malloc(kxp*sizeof(double));
        int* cl2 = (int*) malloc(m*sizeof(int));
        int* nc2 = (int*) malloc(k*sizeof(int));
        double* wss2 = (double*) malloc(k*sizeof(double));
        for (int i = 1; i < nstart; ++i) {
            rand_spl_row(x_unq, mm, p, centers2, k, rand_spl_pool);
            flag_conv = MVR_kmeans_MacQueen(x, centers2,
                                            cl2, nc2, wss2,
                                            m, p, k,
                                            maxiter);
            double z = sum(wss2, k);
            if (z < best) {
                *perror = flag_conv ? 0 : 1;
                best = z;
                
                best_cen = centers2,
                best_cl = cl2,
                best_nc = nc2,
                best_wss = wss2;
            }
        }
        
        if (best_cen != centers) {
            for (int i = 0; i < kxp; ++i) centers[i] = best_cen[i];
            for (int i = 0; i < m; ++i) cl[i] = best_cl[i];
            for (int i = 0; i < k; ++i) nc[i] = best_nc[i];
            for (int i = 0; i < k; ++i) wss[i] = best_wss[i];
        }
        
        best_cen = NULL;
        best_cl = NULL;
        best_nc = NULL;
        best_wss = NULL;
        
        free(centers2);
        free(cl2);
        free(nc2);
        free(wss2);
    }
    
    *tot_wss = best;
    
    free(rand_spl_pool);
    MVR_rand_end();
}


void MVR_withinsumsq(int* pn,
                     int* pp,
                     int* pk,
                     int* pB,
                     double* lWk_bo,
                     int* pnstart,
                     int* pmaxiter,
                     int* perror
                     )
{
    int B = *pB,
        n = *pn,
        p = *pp,
        k = *pk,
        nstart = *pnstart,
        maxiter = *pmaxiter;
    MVR_rand_init();
    double* centers = (double*) malloc(k*p*sizeof(double));
    int* cl = (int*) malloc(n*sizeof(int));
    int* nc = (int*) malloc(k*sizeof(int));
    double* wss = (double*) malloc(k*sizeof(double));
    
    int ref_size = n*p;
    if (ref_size & 1) ++ref_size;
    double* ref = (double*) malloc(ref_size*sizeof(double));
    int* rand_spl_pool = (int*) malloc(n*sizeof(int));
    for (int b = 0; b < B; ++b) {
        rand_norm_std(ref, ref_size);
        double best = R_PosInf;
        for (int i = 0; i < nstart; ++i) {
            rand_spl_row(ref, n, p, centers, k, rand_spl_pool);
            Rboolean flag_conv = MVR_kmeans_MacQueen(ref, centers,
                                                     cl, nc, wss,
                                                     n, p, k,
                                                     maxiter);
            double z = sum(wss, k);
            if (z < best) {
                *perror = flag_conv ? 0 : 1;
                best = z;
            }
        }
        lWk_bo[b] = log(best);
    }
    
    free(centers);
    free(cl);
    free(nc);
    free(wss);
    free(rand_spl_pool);
    free(ref);
    MVR_rand_end();
}


src/MVRc.h
extern "C" {

//=======//
// MVR.h //
//=======//
#ifndef _MVR_H_
#define _MVR_H_

#define __MVR_C_R__
#ifdef __MVR_C_R__
#include <R.h>
#include <Rmath.h>
#else
typedef bool Rboolean;
#define TRUE true
#define FALSE false
#define R_PosInf 1.79769e308
#endif

#endif

//============//
// MVR_stat.h //
//============//
#ifndef _MVR_STAT_H_
#define _MVR_STAT_H_

#include <math.h>
//#include "MVR.h"

double sum(double* a, int n);
double sd(double* a, double mean, int n);
#endif

//============//
// MVR_rand.h //
//============//
#ifndef _MVR_RAND_H_
#define _MVR_RAND_H_

#include <stdlib.h>
#include <time.h>
#include <math.h>
//#include "MVR.h"

//call MVR_rand_init() for once before calling any functions/macros in this library
void MVR_rand_init();
void MVR_rand_end();

/*
RANDI() generates random int number in [0, RANDI_MAX], 30 bits
For 1,000,000 generated numbers, 30 bits has calculated
propensity (probability of 1) from 0.498916 to 0.500959,
and average of 0.5000047333. These values agrees very well
with theoretical values.  Further tests for RANDF()
and RANDD() based on RANDI() confirmed the uniformity test for RANDI()

random float numbers from uniform[0,1.0] distribution
RANDF() has precision of 1e-9
RANDD() has 1e-16 (supposedly 1e-18, but limited by double precision)

uniformity test on generated random numbers by RANDD()
for 1000 sets of data with size of 10,000
Kolmogorov-Smirnov test rejects null hypothesis
for 50 of them at alpha = 0.05
the actual type 1 error of 0.0500 agrees with theoretical well
visually check the histograms for randomly selected 10 sets
all of them seem to follow normality ver well
therefore, the functions passed the normality test :)

RANDF() has lower precision, but still passed the same test 
with type 1 error of 0.0580, slightly bigger than theoretical of 0.05

RANDD() is preferred

Hua XU, PhD
June, 2011
*/
#define RANDI_MAX 0x3fffffff //or 1073741823, 30 bits
//#define RANDI() ((rand()&0x7fff) | ((rand()&0x7fff) << 15))
int randi();

#define RANDF_CONST 1.073741823e9
//#define RANDF() (double(RANDI())/RANDF_CONST)
//double randf();

#define RANDD_CONST1 1.152921504606846975e18
#define RANDD_CONST2 1.073741823999999999e9
//#define RANDD() (double(RANDI())/RANDD_CONST1 + double(RANDI())/RANDD_CONST2)
double randd();

void rand_unif(double* data, int n);

/*
rand_norm_std(data, n)
random numbers from a normal(0,1) distr
From http://www.taygeta.com/random/gaussian.html
Algorithm by Dr. Everett (Skip) Carter, Jr.

Benchmark:
To generate 200 million rand norm numbers
for precision of 1e-9, it takes 27 sec
for precision of 1e-16, it takes 37 sec
R built-in function takes 30 sec
on a computer with Win 7 64bit, Intel(R) Xeon(R) 2.8GHz CPU and 4GB RAM

Normality test on generated random numbers
for 1000 sets of data with size of 10,000
Kolmogorov-Smirnov test rejects null hypothesis
for 48 of them at alpha = 0.05
the actual type 1 error of 0.0480 agrees with theoretical well
visually check the histograms, qqplots for randomly selected 10 sets
all of them seem to follow normality ver well
therefore, the functions passed the normality test :)

rand_norm(data, n, mean, s) passed the same test with type 1 error of 0.0410

Hua XU, PhD
June, 2011
*/
//n must be even for rand_norm and rand_norm_std, OTHERWISE MEMORY OVERFLOWS
void rand_norm_std(double* data, int n);
void rand_norm(double* data, int n, double mean, double s);

/*
Random sampling rows from a matrix
the matrix is mimicked by an array data of size nr x nc as follows:
[0,0] [1,0] [2,0] ... [nr-1,0]
[0,1] ... [nr-1,1]
...
[0,nc-1] ... [nr-1,nc-1]
*/
void rand_spl_row(double* data, int nr, int nc, double* spl, int s, int* pool);
void rand_spl_row2(double* data, int nr, int nc, double* spl, int s);
#endif

//==============//
// MVR_kmeans.h //
//==============//
#ifndef _MVR_KMEANS_H_
#define _MVR_KMEANS_H_

//#include "MVR.h"

/*
More algorithms can be added
*/

Rboolean MVR_kmeans_MacQueen(double *x, double *cen,
                             int *cl, int *nc, double *wss,
                             int n, int p, int k,
                             int maxiter);

#endif

//===========//
// MVR_sub.h //
//===========//
#ifndef _MVR_SUB_H_
#define _MVR_SUB_H_

#include <math.h>
//#include "MVR.h"
//#include "MVR_kmeans.h"
//#include "MVR_stat.h"
//#include "MVR_rand.h"

void MVR_km_clustering(double* x,
                       double* x_unq,
                       
                       double* centers,
                       int* cl,//vec of m
                       int* nc,//vec of k
                       double* wss,//vec of k
                       double* tot_wss,
                       int* perror,
                       
                       int* pm,
                       int* pmm,
                       int* pp,
                       int* pk,
                       int* pnstart,
                       int* pmaxiter);

void MVR_withinsumsq(int* pn,
                     int* pp,
                     int* pk,
                     int* pB,
                     double* lWk_bo,
                     int* pnstart,
                     int* pmaxiter,
                     int* perror
                     );
#endif
}
